Guard `depends_on :xcode` with `if OS.mac?`

jonchang opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Linuxbrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Linuxbrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

https://github.com/Linuxbrew/brew/issues/796

sjackman commented on Formula/dark-mode.rb:
```diff
@@ -13,7 +13,7 @@ class DarkMode < Formula
   depends_on :macos => :el_capitan
-  depends_on :xcode => :build
+  depends_on :xcode => :build if OS.mac?
   depends_on :macos
```
It's not necessary for this formula, which `depends_on :macos`

sjackman commented on Formula/dshb.rb:
```diff
@@ -11,7 +11,7 @@ class Dshb < Formula
 
-  depends_on :xcode => ["8.0", :build]
+  depends_on :xcode => ["8.0", :build] if OS.mac?
   depends_on :macos
```
Ditto.

sjackman commented on Formula/fmdiff.rb:
```diff
@@ -17,7 +17,7 @@ class Fmdiff < Formula
   # Prior to that it was included in the Developer Tools package.
   # "make" has logic for checking both possibilities.
-  depends_on :xcode if MacOS.version >= :lion
+  depends_on :xcode if MacOS.version >= :lion if OS.mac?
```
`brew audit` will complain about two `if` on the same line.
```ruby
depends_on :xcode if OS.mac? && MacOS.version >= :lion
```

sjackman commented on Formula/freerdp.rb:
```diff
@@ -34,7 +34,7 @@ class Freerdp < Formula
     sha256 "1a4b03daca9bd82ef2e3827e1cdb469121f62aab1061f79ffa587babf9fe080c"
     version "1.1.0-beta1"
-    depends_on :xcode => :build # for "ibtool"
+    depends_on :xcode => :build # for "ibtool" if OS.mac?
```
It's after a comment here.

sjackman commented on Formula/graphviz.rb:
```diff
@@ -33,7 +33,7 @@ class Graphviz < Formula
 
   depends_on "pkg-config" => :build
-  depends_on :xcode => :build if build.with? "app"
+  depends_on :xcode => :build if build.with? "app" if OS.mac?
```
Two ifs.

sjackman commented on Formula/libmxml.rb:
```diff
@@ -13,7 +13,7 @@ class Libmxml < Formula
   end
 
-  depends_on :xcode => :build # for docsetutil
+  depends_on :xcode => :build # for docsetutil if OS.mac?
```
Comment

sjackman commented on Formula/mplayershell.rb:
```diff
@@ -17,7 +17,7 @@ class Mplayershell < Formula
   depends_on :macos => :lion
-  depends_on :xcode => :build
+  depends_on :xcode => :build if OS.mac?
   depends_on :macos
```
macOS-only

sjackman commented on Formula/pngpaste.rb:
```diff
@@ -14,7 +14,7 @@ class Pngpaste < Formula
 
   # Sierra's CLT is sufficient, but El Capitain's isn't
-  depends_on :xcode => [:build, "8.0"] if MacOS.version < :sierra
+  depends_on :xcode => [:build, "8.0"] if MacOS.version < :sierra and OS.mac?
```
`&&` rather than `and`
```ruby
depends_on :xcode => [:build, "8.0"] if OS.mac? && MacOS.version < :sierra
```

jonchang commented:
Even if a formula has `depends_on :macos`, `brew info FORMULA` will still fail. E.g.

```
 $ brew info dark-mode
dark-mode: stable 2.0.1, HEAD
Control the macOS dark mode from the command-line
https://github.com/sindresorhus/dark-mode
Not installed
From: https://github.com/Linuxbrew/homebrew-core/blob/master/Formula/dark-mode.rb
==> Requirements
Error: undefined method `installed?' for OS::Mac::Xcode:Module
Did you mean?  instance_of?
```

sjackman commented on Formula/sourcekitten.rb:
```diff
@@ -13,8 +13,8 @@ class Sourcekitten < Formula
-  depends_on :xcode => "6.0"
-  depends_on :xcode => ["9.0", :build]
+  depends_on :xcode => "6.0" if OS.mac?
+  depends_on :xcode => ["9.0", :build] if OS.mac?
```
macOS-only

sjackman commented on Formula/sourcery.rb:
```diff
@@ -10,8 +10,8 @@ class Sourcery < Formula
-  depends_on :xcode => ["9.3", :build]
+  depends_on :xcode => "6.0" if OS.mac?
+  depends_on :xcode => ["9.3", :build] if OS.mac?
 
```
```ruby
if OS.mac?
…
end
```

sjackman commented on Formula/swiftlint.rb:
```diff
@@ -15,8 +15,8 @@ class Swiftlint < Formula
-  depends_on :xcode => "8.0"
-  depends_on :xcode => ["9.0", :build]
+  depends_on :xcode => "8.0" if OS.mac?
+  depends_on :xcode => ["9.0", :build] if OS.mac?
```
Use an if block

sjackman commented on Formula/xctool.rb:
```diff
@@ -12,7 +12,7 @@ class Xctool < Formula
 
-  depends_on :xcode => "7.0"
+  depends_on :xcode => "7.0" if OS.mac?
   depends_on :macos
```
macOS-only

sjackman commented:
Thanks, Joe! I've made a few comments.
What error do you see with Homebrew/brew if a formula has
```ruby
depends_on :xcode
```
?

jonchang commented:
```
 $ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

 $ brew info xctool
xctool: stable 0.3.4, HEAD
Drop-in replacement for xcodebuild with a few extra features
https://github.com/facebook/xctool
Not installed
From: https://github.com/Linuxbrew/homebrew-core/blob/master/Formula/xctool.rb
==> Requirements
Error: undefined method `installed?' for OS::Mac::Xcode:Module
Did you mean?  instance_of?
Please report this bug:
  https://github.com/Linuxbrew/brew/wiki/troubleshooting
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirements/xcode_requirement.rb:14:in `xcode_installed_version'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirements/xcode_requirement.rb:6:in `block in <class:XcodeRequirement>'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirement.rb:56:in `instance_eval'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirement.rb:56:in `block in satisfied?'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirement.rb:186:in `yielder'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirement.rb:56:in `satisfied?'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:198:in `block in decorate_requirements'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:196:in `collect'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:196:in `decorate_requirements'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:171:in `block in info_formula'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:168:in `map'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:168:in `info_formula'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:58:in `block in print_info'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:52:in `each'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:52:in `each_with_index'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:52:in `print_info'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:41:in `info'
/home/linuxbrew/.linuxbrew/Library/Homebrew/brew.rb:87:in `<main>'

 $ git checkout guard_xcode
Switched to branch 'guard_xcode'
Your branch is up to date with 'jonchang/guard_xcode'.


 $ brew info xctool
xctool: stable 0.3.4, HEAD
Drop-in replacement for xcodebuild with a few extra features
https://github.com/facebook/xctool
Not installed
From: https://github.com/Linuxbrew/homebrew-core/blob/master/Formula/xctool.rb
==> Requirements
Required: macOS is required (uninstalled)
==> Options
--HEAD
	Install HEAD version
```

jonchang commented:
I believe this is because `brew info` will look at all of the `depends_on`, and doing `depends_on :xcode` on non-macOS will :boom:

jonchang commented:
In linuxbrew/brew this is addressed by just doing `return true` in `XcodeRequirement` https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/requirements/xcode_requirement.rb#L14

sjackman commented:
Ah, got it

sjackman commented:
Thanks for this work, Jon!
